sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala

```diff
@@ -194,6 +258,9 @@ case class PlanSubqueries(sparkSession: SparkSession) extends Rule[SparkPlan] {
         }
         val executedPlan = new QueryExecution(sparkSession, query).executedPlan
         InSubqueryExec(expr, SubqueryExec(s"subquery#${exprId.id}", executedPlan), exprId)
+      case expressions.Exists(sub, children, exprId) =>
+        val executedPlan = new QueryExecution(sparkSession, Project(Nil, sub)).executedPlan
```

Comment:
Can we add the Project a bit earlier? This seems weird to add Project and do column pruning at planning time.

Reply:
> Can we add the Project a bit earlier? This seems weird to add Project and do column pruning at planning time.

add this in `RewritePredicateSubquery`?

Reply:
> Can we add the Project a bit earlier? This seems weird to add Project and do column pruning at planning time.

add this in `RewritePredicateSubquery`?

Reply:
how about the rule `OptimizeSubqueries`?

Reply:
> how about the rule `OptimizeSubqueries`?

correlated and non-correlated subqueries all  match here, and we only handle non-correlated exists.  If we add `Project(Nil, ..)` to here need to check subquery type.
For me, it ok to add it to here since it just handle one case and we can add comment here to explain why we do this.

Reply:
Isn't it useful for correlated EXISTS as well? logically it's always valid to prune all columns from EXISTS subquery, no matter it's correlated or not.

Reply:
> Isn't it useful for correlated EXISTS as well? logically it's always valid to prune all columns from EXISTS subquery, no matter it's correlated or not.

For correlated exists,  it will be changed to a `join`, add `Project(Nil, ...)` in `OptimizeSubqueries ` will break the column reference.

Reply:
I checked `PullupCorrelatedPredicates`, and we will add the missing references.

I don't think `Project(Nil, ...)` can cause problems. People can always write queries with missing references, e.g. `EXISTS (SELECT a FROM t WHERE t.b = outer_table.col)`, and the framework must be prepared for it.

Reply:
> I checked `PullupCorrelatedPredicates`, and we will add the missing references.
> 
> I don't think `Project(Nil, ...)` can cause problems. People can always write queries with missing references, e.g. `EXISTS (SELECT a FROM t WHERE t.b = outer_table.col)`, and the framework must be prepared for it.

With UT below, failed.
```
 test("SPARK-29800") {
    withTempView("s1", "s2", "s3") {
      Seq(1, 3, 5, 7, 9).toDF("id").createOrReplaceTempView("s1")
      Seq(1, 3, 4, 6, 9).toDF("id").createOrReplaceTempView("s2")
      Seq(3, 4, 6, 9).toDF("id").createOrReplaceTempView("s3")
      sql(
        """
          | SELECT s1.id FROM s1
          | JOIN s2 ON s1.id = s2.id
          | WHERE EXISTS (SELECT * from s3 where s3.id = s1.id)
        """.stripMargin).show()
     }
}
```

Reply:
Lose reference `id#256`
```
== Parsed Logical Plan ==
'Project ['s1.id]
+- 'Filter exists#258 []
   :  +- 'Project [*]
   :     +- 'Filter ('s3.id = 's1.id)
   :        +- 'UnresolvedRelation [s3]
   +- 'Join Inner, ('s1.id = 's2.id)
      :- 'UnresolvedRelation [s1]
      +- 'UnresolvedRelation [s2]

== Analyzed Logical Plan ==
id: int
Project [id#244]
+- Filter exists#258 [id#244]
   :  +- Project [id#256]
   :     +- Filter (id#256 = outer(id#244))
   :        +- SubqueryAlias `s3`
   :           +- Project [value#253 AS id#256]
   :              +- LocalRelation [value#253]
   +- Join Inner, (id#244 = id#250)
      :- SubqueryAlias `s1`
      :  +- Project [value#241 AS id#244]
      :     +- LocalRelation [value#241]
      +- SubqueryAlias `s2`
         +- Project [value#247 AS id#250]
            +- LocalRelation [value#247]

== Optimized Logical Plan ==
!Join LeftSemi, (id#256 = id#244)
:- Project [id#244]
:  +- Join Inner, (id#244 = id#250)
:     :- Project [value#241 AS id#244]
:     :  +- LocalRelation [value#241]
:     +- Project [value#247 AS id#250]
:        +- LocalRelation [value#247]
+- Project
   +- LocalRelation [value#253]

== Physical Plan ==
!BroadcastNestedLoopJoin BuildRight, LeftSemi, (id#256 = id#244)
:- *(2) Project [id#244]
:  +- *(2) BroadcastHashJoin [id#244], [id#250], Inner, BuildRight
:     :- *(2) Project [value#241 AS id#244]
:     :  +- *(2) LocalTableScan [value#241]
:     +- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, false] as bigint))), [id=#129]
:        +- *(1) Project [value#247 AS id#250]
:           +- *(1) LocalTableScan [value#247]
+- BroadcastExchange IdentityBroadcastMode, [id=#135]
   +- *(3) Project
      +- *(3) LocalTableScan [value#253]
```

Reply:
how about WHERE? it's a known problem that correlated subqueries may not work well in join condition.

Reply:
> how about WHERE on a single table? it's a known problem that correlated subqueries may not work well in join condition.

It is ok with WHERE on a single table.
The case I show you is WHERE with correlated subquery, failed, not in join condition

Reply:
For correlated exists, convert to join, seems don't need to do this?

Reply:
OK let's keep it here now. But I do think it's better to put logical optimization in logical rules.

Reply:
> OK let's keep it here now. But I do think it's better to put logical optimization in logical rules.

Understand, but current code about subqueries disperse in different places and interacting with each other